Sources/Web3Modal/Screens/ChainSwitch/NetworkDetail/NetworkDetailViewModel.swift

```diff
@@ -0,0 +1,218 @@
+    enum Event {
+        case onAppear
+        case didTapRetry
+    }
+    
+    @Published var switchFailed: Bool = false
+    var triedAddingChain: Bool = false
+    
+    let chain: Chain
+    let router: Router
+    let store: Store
+    
+    private var disposeBag = Set<AnyCancellable>()
+    
+    init(
+        chain: Chain,
+        router: Router,
+        store: Store = .shared
+    ) {
+        self.chain = chain
+        self.router = router
+        self.store = store
+        
+        Web3Modal.instance.sessionEventPublisher
+            .receive(on: DispatchQueue.main)
+            .sink { (event: Session.Event, _: String, _: Blockchain?) in
+                if event.name == "chainChanged" {
+                    
+                    guard let chainReference = try? event.data.get(Int.self) else {
+                        return
+                    }
+                    
+                    DispatchQueue.main.async {
+                        self.store.selectedChain = ChainPresets.ethChains.first(where: { $0.chainReference == String(chainReference) })
+                        self.router.setRoute(Router.AccountSubpage.profile)
+                    }
+                }
+            }
+            .store(in: &disposeBag)
+        
+        Web3Modal.instance.sessionResponsePublisher
+            .receive(on: DispatchQueue.main)
+            .sink { response in
+                switch response.result {
+                case let .response(value):
+                    let stringResponse = try? value.get(String.self)
+                    
+                    DispatchQueue.main.async {
+                        self.store.selectedChain = chain
+                        self.router.setRoute(Router.AccountSubpage.profile)
+                    }
+                case let .error(error):
+                    
+                    if !self.triedAddingChain {
+                        guard let from = store.selectedChain else {
+                            return
+                        }
+                        
+                        self.triedAddingChain = true
+                        
+                        Task {
+                            try? await self.addEthChain(from: from, to: chain)
+                        }
+                    } else {
+                        DispatchQueue.main.async {
+                            self.switchFailed = true
+                            self.objectWillChange.send()
+                        }
+                    }
+                }
+            }
+        
+            .store(in: &disposeBag)
+    }
+    
+    func handle(_ event: Event) {
+        switch event {
+        case .onAppear:
+            Task { @MainActor in
+                // Switch chain
+                await switchChain(chain)
+            }
+        case .didTapRetry:
+            
+            self.triedAddingChain = false
+            self.switchFailed = false
+            Task { @MainActor in
```

Comment:
I'm not sure that switchChain will be executed on main thread. You should mark switchChain method with @MainActor ig